pbparam/optimisation_problems/data_fit.py

```diff
@@ -29,16 +29,25 @@ def update_simulation_parameters(simulation, parameter_values):
 
     return new_simulation
 
-def cost_function_full(self, simulation, map_inputs, scalings, data, variables_optimise, x):
+def cost_function_full(opt_problem, x):
+
     # TODO: allow for multifunction optimisation, and for various cost functions
+
+    simulation = opt_problem.simulation
+    map_inputs = opt_problem.map_inputs
+    scalings = opt_problem.scalings
+    data = opt_problem.data
+    variables_optimise = opt_problem.variables_optimise
+    
     input_dict = {param: scalings[i] * x[i] for param, i in map_inputs.items()}
     t_end = data["Time [s]"].iloc[-1]
     solution = simulation.solve([0, t_end], inputs=input_dict)
     norm_tot_func = 0
+    cost_function = copy.deepcopy(cost_function)
```

Comment:
Why do you need the `deepcopy`? Does it not work if you simply call `cost_function.evaluate`?